### DISCUSSION TITLE
ManytoMany field on prefix model

### DISCUSSION THREAD

**Discussion by andy-shady-org** [netbox-community/netbox#10551] [General]
Hi,

When I build a many to many relationship on Prefix model, I get the following error:

... "X needs to have a value for field "id" before this many-to-many relationship can be used."

However, when I build a many to many relationship on IPAddress or VRF models, I dont get this error.

Works:

    source_zones = models.ManyToManyField(
        VRF,
        blank=True,
        related_name='rule_source_zones'
    )

    source_addresses = models.ManyToManyField(
        IPAddress,
        blank=True,
        related_name='rule_source_addresses'
    )

Does not work:

    source_prefixes = models.ManyToManyField(
        Prefix,
        blank=True,
        related_name='rule_source_prefixes',
    )

Is there any reason for this?

Thanks

Andy

**Comment by jeremystretch** [comment]
> "X needs to have a value for field "id" before this many-to-many relationship can be used."

This indicates that you're trying to save a many-to-many relationship between objects before one of the objects itself has been saved. Have a read through [this page](https://docs.djangoproject.com/en/4.1/topics/db/examples/many_to_many/) in the Django docs for some examples of how to work with M2M relationships.

**Reply by andy-shady-org** [reply]
Jeremy,

I read through that, and saw the alternate option to use ParentalManytoMany field also.
If I take a step back from that, The field object is created, as its a form selectable prefix. 
That leaves only the parent model to be saved, when all related field objects are already created.

This works perfectly fine for IPAddress, and VRF fields, both of them are created prior to the creation of the parent model, selectable within forms, showing they have been created, etc.

However, this does not work for Prefix, when the Prefix object has been pre-created, and like IPaddress and VRF objects, is selected from a form.

Would you suggest overloading the save() method of the model class to ensure that the 'id' exists, then saving the model again?
If not, how would you suggest calling save without the field being set, and then setting it and calling add() or save() again? Would this would be done within a form or serialiser?

Thanks

**Reply by andy-shady-org** [reply]
fixed it, the issue was in my clean method which was doing some pre-checks.
